### ISSUE TITLE
[HOLD App #28021] Scan - Unable to delete receipt after replacing it unless merchant name is entered

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#29766]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.3.85-0

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

**Issue found when executing PR** https://github.com/Expensify/App/pull/29617
## Action Performed:
1. Go to staging.new.expensify.com
2. Go to + > Request money > Scan
3. Complete the scan request
4. Open Scan request details page
5. Click on the receipt > 3-dot menu > Replace
6. Upload a photo
7. Click on the receipt - 3-dot menu > Delete receipt > Delete
8. Enter merchant name
9. Click on the receipt - 3-dot menu > Delete receipt > Delete

## Expected Result:
In Step 7, if the receipt cannot be deleted for some reason, there should be an error message or the delete action button should be hidden
In Step 9, if the empty fields must be filled before the receipt can be deleted, the app should only allow the deletion of receipt after both Amount and Merchant values are entered

## Actual Result:
In Step 7, after deleting the receipt, it disappears and then reappears
In Step 9, the receipt can only be deleted after Merchant name is entered, while the amount field is still empty

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

 

https://github.com/Expensify/App/assets/78819774/e8e93806-8fac-4ae7-b295-2dbcc1a203b1

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by paultsimura** [comment]
If we want to maintain the current behavior and just allow removing the receipt, the changes should be made on the BE side, as the `DetachReceipt` API request returns the following:

```json
{
    "jsonCode": 405,
    "message": "405 Cannot detach receipt from partial transaction",
    "onyxData": [],
    "requestID": "8179546a49d234ee-WAW"
}
```

I suppose the receipt state should be taken into consideration on the BE: if it's not scanning – then allow detaching.
Otherwise, it'd be better to clarify the expected behavior: hide the delete button, update the merchant automatically on the receipt replacement, etc.

**Comment by trjExpensify** [comment]
@Gonals @luacmartins can you look at this one please? 

I'm not 100% sure it'll be fixed in @joekaufmanexpensify [issue](https://github.com/Expensify/App/issues/28021).

**Comment by paultsimura** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The scan report cannot be removed after the receipt is replaced.

### What is the root cause of that problem?
The root cause lies here:

https://github.com/Expensify/App/blob/c946ba77754936bcbd22e673772bfc0b838a6bc6/src/pages/iou/ReceiptSelector/index.js#L332-L337

In the ReceiptSelector, we are trying to get the `reportID` from the path. This works for the new requests, but if we are replacing the receipt, the route params look like this:

```json
{
    "key": "EditRequest_Root-xiCzvsFIN5uLz1knTTGpu",
    "name": "EditRequest_Root",
    "params": {
        "threadReportID": "5081786922064036",
        "field": "receipt"
    }
}
```

So the report ID here is `threadReportID`, not `reportID`.

### What changes do you think we should make in order to solve the problem?
We should fetch the report by either reportID or, if absent, - try the threadReportID:

```js
export default withOnyx({
    iou: {key: ONYXKEYS.IOU},
    report: {
        key: ({route}) => {
            const reportID = lodashGet(route, 'params.reportID') || lodashGet(route, 'params.threadReportID', '');
            return `${ONYXKEYS.COLLECTION.REPORT}${reportID}`;
        },
    },
})(ReceiptSelector);
```

This will allow us to handle both cases: new IOU report creation and receipt replacement.



### What alternative solutions did you explore? (Optional)

**Comment by trjExpensify** [comment]
Bump @Gonals @luacmartins on the above! https://github.com/Expensify/App/issues/29766#issuecomment-1767332021

**Comment by luacmartins** [comment]
I agree with @paultsimura, this seems to be an internal issue

**Comment by trjExpensify** [comment]
Okay, but that wasn't really my question. 😅 The issue linked to has a similar root cause in that you are able to replace a receipt while it's still scanning. After which, you can run into these downstream "bugs" as side effects; 1) you can't delete the receipt 2) other participants can't see the replaced receipt

So just like we hide the ability to detach a receipt while it's still scanning, we should hide the ability to replace a receipt while it's still scanning. That avoids both of these issues. So my question is if that's going to be implemented in #28021? If so, can this issue go on hold until a PR is merged for that as we don't need to do anything in addition whether that be on the BE or FE.

**Comment by luacmartins** [comment]
Ah got it. It seems like we're leaning towards hiding the replace option while scan is in progress, but we still don't have a proposal for https://github.com/Expensify/App/issues/28021. That being said, I agree that we can put this issue on hold until we have a PR for that issue. Thanks for raising this!

**Comment by trjExpensify** [comment]
Cool, cool. Dropping to weekly while it's on hold.

**Comment by trjExpensify** [comment]
Samesies.

**Comment by trjExpensify** [comment]
Still held.

**Comment by trjExpensify** [comment]
Same spot.

**Comment by trjExpensify** [comment]
Still held. Waiting on @Gonals to provide a secondary review of the proposal in: https://github.com/Expensify/App/issues/28021#issuecomment-1847325586

**Comment by trjExpensify** [comment]
Still held, PR is in review.

**Comment by trjExpensify** [comment]
PR hit staging yesterday, and I can see you can't run into this issue anymore because the ability to replace a receipt while it's still scanning has been removed. Closing!
